x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java

```diff
@@ -0,0 +1,297 @@
+     */
+    public void listEngine(String queryString, int from, int size, ActionListener<SearchResponse> listener) {
+        try {
+            final SearchSourceBuilder source = new SearchSourceBuilder().from(from)
+                .size(size)
+                .query(new QueryStringQueryBuilder(queryString))
+                .docValueField(NAME_FIELD.getPreferredName())
+                .docValueField(INDICES_FIELD.getPreferredName())
+                .storedFields(Collections.singletonList("_none_"))
+                .sort(NAME_FIELD.getPreferredName(), SortOrder.ASC);
+            final SearchRequest req = new SearchRequest(ENGINE_ALIAS_NAME).source(source);
+            clientWithOrigin.search(req, listener);
+        } catch (Exception e) {
+            listener.onFailure(e);
+        }
+    }
+
+    private Engine parseEngineBinaryFromSource(BytesReference source) {
+        try (XContentParser parser = XContentHelper.createParser(XContentParserConfiguration.EMPTY, source, XContentType.JSON)) {
+            ensureExpectedToken(parser.nextToken(), XContentParser.Token.START_OBJECT, parser);
+            while (parser.nextToken() != XContentParser.Token.END_OBJECT) {
+                ensureExpectedToken(XContentParser.Token.FIELD_NAME, parser.currentToken(), parser);
+                parser.nextToken();
+                if (BINARY_CONTENT_FIELD.getPreferredName().equals(parser.currentName())) {
+                    final CharBuffer encodedBuffer = parser.charBuffer();
+                    InputStream encodedIn = Base64.getDecoder().wrap(new InputStream() {
+                        @Override
+                        public int read() {
+                            if (encodedBuffer.hasRemaining()) {
+                                return encodedBuffer.get();
+                            } else {
+                                return -1; // end of stream
+                            }
+                        }
+                    });
+                    try (
+                        StreamInput in = new NamedWriteableAwareStreamInput(new InputStreamStreamInput(encodedIn), namedWriteableRegistry)
+                    ) {
+                        return parseEngineBinaryWithVersion(in);
+                    }
+                } else {
+                    XContentParserUtils.parseFieldsValue(parser); // consume and discard unknown fields
+                }
+            }
+            throw new ElasticsearchParseException("[" + BINARY_CONTENT_FIELD.getPreferredName() + "] field is missing");
+        } catch (IOException e) {
+            throw new ElasticsearchParseException("Failed to parse: " + source.utf8ToString(), e);
+        }
+    }
+
+    static Engine parseEngineBinaryWithVersion(StreamInput in) throws IOException {
+        TransportVersion version = TransportVersion.readVersion(in);
+        assert version.onOrBefore(TransportVersion.CURRENT) : version + " >= " + TransportVersion.CURRENT;
```

Comment:
I'm confused about why we need to do this serialization. Can't this assertion be achieved by adding a `version` field?

Reply:
Yes it could. Let's discuss this offline and we can change the internals later.